File: test-mqtt.js
const mqtt = require('mqtt');
require('dotenv').config();

console.log('🧪 [MQTT_TEST] Iniciando prueba de conectividad MQTT...');

// Configuración MQTT desde variables de entorno
const mqttConfig = {
    hostname: process.env.MQTT_BROKER_URL?.replace('mqtt://', '').split(':')[0] || '172.30.1.21',
    port: parseInt(process.env.MQTT_BROKER_URL?.split(':')[2]) || 1883,
    username: process.env.MQTT_USERNAME || 'mqttclient',
    password: process.env.MQTT_PASSWORD || 'arkus@123',
    topic: process.env.MQTT_TOPIC || 'arkus/n1/switch/door_relay'
};

console.log('📊 [MQTT_TEST] Configuración:');
console.log(`   - Broker: ${mqttConfig.hostname}:${mqttConfig.port}`);
console.log(`   - Usuario: ${mqttConfig.username}`);
console.log(`   - Tópico: ${mqttConfig.topic}`);

// Conectar al broker MQTT
const client = mqtt.connect(`mqtt://${mqttConfig.hostname}:${mqttConfig.port}`, {
    username: mqttConfig.username,
    password: mqttConfig.password,
    reconnectPeriod: 1000,
    connectTimeout: 10000
});

let testCompleted = false;

client.on('connect', () => {
    console.log('✅ [MQTT_TEST] Conectado exitosamente al broker MQTT');
    
    // Suscribirse al tópico para escuchar mensajes
    client.subscribe(mqttConfig.topic, (error) => {
        if (error) {
            console.error('❌ [MQTT_TEST] Error suscribiéndose al tópico:', error);
        } else {
            console.log(`📡 [MQTT_TEST] Suscrito al tópico: ${mqttConfig.topic}`);
        }
    });
    
    // Enviar mensaje de prueba
    setTimeout(() => {
        sendTestMessage();
    }, 1000);
});

client.on('message', (topic, message) => {
    console.log(`📨 [MQTT_TEST] Mensaje recibido en ${topic}:`);
    console.log(`   Contenido: ${message.toString()}`);
    
    if (!testCompleted) {
        testCompleted = true;
        console.log('✅ [MQTT_TEST] Prueba completada exitosamente');
        client.end();
    }
});

client.on('error', (error) => {
    console.error('❌ [MQTT_TEST] Error de conexión MQTT:', error);
    process.exit(1);
});

client.on('close', () => {
    console.log('🔌 [MQTT_TEST] Conexión MQTT cerrada');
    if (!testCompleted) {
        console.log('❌ [MQTT_TEST] Prueba fallida - conexión cerrada inesperadamente');
        process.exit(1);
    }
});

function sendTestMessage() {
    const testMessage = JSON.stringify({
        action: 'test',
        user: 'Test User',
        timestamp: new Date().toISOString(),
        source: 'opendoor-test'
    });
    
    console.log('📤 [MQTT_TEST] Enviando mensaje de prueba...');
    console.log(`   Mensaje: ${testMessage}`);
    
    client.publish(mqttConfig.topic, testMessage, (error) => {
        if (error) {
            console.error('❌ [MQTT_TEST] Error enviando mensaje:', error);
            process.exit(1);
        } else {
            console.log('✅ [MQTT_TEST] Mensaje enviado exitosamente');
            console.log('⏳ [MQTT_TEST] Esperando respuesta...');
            
            // Timeout después de 10 segundos
            setTimeout(() => {
                if (!testCompleted) {
                    console.log('⏰ [MQTT_TEST] Timeout - no se recibió respuesta');
                    console.log('💡 [MQTT_TEST] Esto puede ser normal si no hay otros clientes suscritos');
                    testCompleted = true;
                    client.end();
                }
            }, 10000);
        }
    });
}

// Manejar señales de terminación
process.on('SIGINT', () => {
    console.log('\n🛑 [MQTT_TEST] Recibida señal SIGINT, terminando...');
    client.end();
    process.exit(0);
});

process.on('SIGTERM', () => {
    console.log('\n🛑 [MQTT_TEST] Recibida señal SIGTERM, terminando...');
    client.end();
    process.exit(0);
});

// Timeout general de 30 segundos
setTimeout(() => {
    if (!testCompleted) {
        console.log('⏰ [MQTT_TEST] Timeout general - terminando prueba');
        client.end();
        process.exit(0);
    }
}, 30000);

File: test_mqtt.py
#!/usr/bin/env python3
"""
Script de prueba para verificar la conexión MQTT
"""

import paho.mqtt.client as mqtt
import time
import os
from dotenv import load_dotenv

# Cargar variables de entorno
load_dotenv()

# Configuración MQTT
MQTT_BROKER_URL = os.getenv('MQTT_BROKER_URL', 'localhost')
MQTT_PORT = int(os.getenv('MQTT_PORT', '1883'))
MQTT_USERNAME = os.getenv('MQTT_USERNAME', 'mqttclient')
MQTT_PASSWORD = os.getenv('MQTT_PASSWORD', 'arkus@123')
MQTT_TOPIC = os.getenv('MQTT_TOPIC', 'arkus/n1/switch/door_relay')

# Limpiar la URL del broker si incluye protocolo
if MQTT_BROKER_URL.startswith('mqtt://'):
    MQTT_BROKER_URL = MQTT_BROKER_URL.replace('mqtt://', '')
elif MQTT_BROKER_URL.startswith('tcp://'):
    MQTT_BROKER_URL = MQTT_BROKER_URL.replace('tcp://', '')

def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("✅ [MQTT] Conectado al broker MQTT exitosamente")
        print(f"   📍 [MQTT] Broker: {MQTT_BROKER_URL}:{MQTT_PORT}")
        print(f"   🔐 [MQTT] Usuario: {MQTT_USERNAME}")
        print(f"   📡 [MQTT] Tópico: {MQTT_TOPIC}")
    else:
        print(f"❌ [MQTT] Error de conexión MQTT: {rc}")
        print("   💡 [MQTT] Códigos de error:")
        print("      - 1: Protocolo incorrecto")
        print("      - 2: Identificador de cliente inválido")
        print("      - 3: Servidor no disponible")
        print("      - 4: Usuario o contraseña incorrectos")
        print("      - 5: No autorizado")

def on_publish(client, userdata, mid):
    print(f"✅ [MQTT] Mensaje publicado exitosamente (ID: {mid})")

def on_disconnect(client, userdata, rc):
    if rc != 0:
        print(f"⚠️ [MQTT] Desconexión inesperada (código: {rc})")
    else:
        print("✅ [MQTT] Desconexión normal")

def test_mqtt_connection():
    print("🧪 [TEST] Iniciando prueba de conexión MQTT...")
    print("=" * 60)
    
    # Crear cliente MQTT
    client = mqtt.Client()
    
    # Configurar callbacks
    client.on_connect = on_connect
    client.on_publish = on_publish
    client.on_disconnect = on_disconnect
    
    try:
        # Configurar autenticación
        if MQTT_USERNAME and MQTT_PASSWORD:
            client.username_pw_set(MQTT_USERNAME, MQTT_PASSWORD)
            print(f"🔐 [MQTT] Autenticación configurada: {MQTT_USERNAME}")
        
        # Conectar
        print(f"🔄 [MQTT] Conectando a {MQTT_BROKER_URL}:{MQTT_PORT}...")
        client.connect(MQTT_BROKER_URL, MQTT_PORT, 60)
        
        # Iniciar loop
        client.loop_start()
        
        # Esperar conexión
        time.sleep(2)
        
        if client.is_connected():
            print("✅ [MQTT] Cliente conectado y funcionando")
            
            # Probar publicación
            print(f"📤 [MQTT] Probando publicación en tópico: {MQTT_TOPIC}")
            result = client.publish(MQTT_TOPIC, "TEST_MESSAGE", qos=1)
            
            if result.rc == mqtt.MQTT_ERR_SUCCESS:
                print("✅ [MQTT] Publicación de prueba exitosa")
            else:
                print(f"❌ [MQTT] Error en publicación de prueba: {result.rc}")
            
            # Esperar un poco más
            time.sleep(2)
            
            # Desconectar
            client.loop_stop()
            client.disconnect()
            print("✅ [MQTT] Cliente desconectado")
            
        else:
            print("❌ [MQTT] No se pudo establecer conexión")
            
    except Exception as e:
        print(f"❌ [MQTT] Error durante la prueba: {e}")
        print(f"   📍 [MQTT] Broker: {MQTT_BROKER_URL}")
        print(f"   📍 [MQTT] Puerto: {MQTT_PORT}")
        
        # Sugerencias de solución
        print("\n💡 [SOLUCIONES] Posibles soluciones:")
        print("   1. Verifica que Mosquitto esté corriendo:")
        print("      - Windows: sc query mosquitto")
        print("      - Linux: systemctl status mosquitto")
        print("   2. Verifica el puerto:")
        print("      - netstat -an | findstr 1883")
        print("   3. Verifica la configuración del broker:")
        print("      - Archivo mosquitto.conf")
        print("      - Usuarios y contraseñas")
        print("   4. Prueba con mosquitto_sub:")
        print("      - mosquitto_sub -h localhost -p 1883 -t 'test/topic'")

if __name__ == "__main__":
    test_mqtt_connection()
